src/anomalib/pipelines/tiled_ensemble/train_pipeline.py

```diff
@@ -0,0 +1,74 @@
+"""Tiled ensemble training pipeline."""
+
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from pathlib import Path
+
+import logging
+
+import torch
+
+from anomalib.pipelines.components.base import Pipeline, Runner
+from anomalib.pipelines.components.runners import ParallelRunner, SerialRunner
+
+from .components import (
+    MergeJobGenerator,
+    PredictJobGenerator,
+    SmoothingJobGenerator,
+    StatisticsJobGenerator,
+    TrainModelJobGenerator,
+)
+from .components.utils import PredictData
+from .components.utils.ensemble_engine import TiledEnsembleEngine
+
+logger = logging.getLogger(__name__)
+
+
+class TrainTiledEnsemble(Pipeline):
+    """Tiled ensemble training pipeline."""
+
+    def __init__(self) -> None:
+        self.root_dir: Path | None = None
+
+    def _setup_runners(self, args: dict) -> list[Runner]:
+        """Setup the runners for the pipeline.
+
+        This pipeline consists of training and validation steps:
+        Training models > prediction on val data > merging val data >
+        > (optionally) smoothing seams > calculation of post-processing statistics
+
+        Returns:
+            list[Runner]: List of runners executing tiled ensemble train + val jobs.
+        """
+        runners: list[Runner] = []
+        self.root_dir = TiledEnsembleEngine.setup_ensemble_workspace(args["pipeline"])
+
+        if args["pipeline"]["accelerator"] == "cuda":
+            runners.extend(
+                [
+                    ParallelRunner(TrainModelJobGenerator(self.root_dir), n_jobs=torch.cuda.device_count()),
+                    ParallelRunner(
+                        PredictJobGenerator(self.root_dir, PredictData.VAL),
+                        n_jobs=torch.cuda.device_count(),
+                    ),
+                ],
+            )
+        else:
+            runners.extend(
+                [
+                    SerialRunner(TrainModelJobGenerator(self.root_dir)),
+                    SerialRunner(PredictJobGenerator(self.root_dir, PredictData.VAL)),
+                ],
+            )
+        runners.append(SerialRunner(MergeJobGenerator()))
+
+        if args["pipeline"]["ensemble"]["post_processing"]["seam_smoothing"]["apply"]:
```

Comment:
The solution above should work for this case and is much nicer than what I have now. 

The problem here is that some pipeline stages require args from some other part of config. For example, model training job needs to know normalization stage to determine if normalization should be applied at tile level, but normalization stage is is part of post processing. Since the pipeline class does this `_args = args.get(runner.generator.job_class.name, None)` it would then mean that train job doesn't have access, or I'd need to duplicate this.

So I'm not sure how to best handle such cases with current design.

Reply:
One possibility is to somehow accumulate the ones that are shared under one name, but then the config file looses the structure a bit.